### PR TITLE
Add http, bash, jp, jq, ps

### PR DESCRIPTION
**Pull request by itsofirblink**

### CODE REVIEW

File: actions/http_delete.action.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Describes the action and it's parameters
```
Comment (rimar-blink) [inline actions/http_delete.action.yaml:1]: Why this isn't just 1 http action with method parameter?

File: actions/http_get.action.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Describes the action and it's parameters
```
Comment (rimar-blink) [inline actions/http_get.action.yaml:1]: need to be able to add http headers

File: actions/http_post.action.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    description: "The url to send a post request"
+    required: true
+  contentType:
+    type: "string"
+    description: "Content type for post request"
+    required: true
+  body:
+    type: "string"
+    description: "Body for post request"
+    required: true
```
Comment (rimar-blink) [inline actions/http_post.action.yaml:17]: body can be empty - will it work with the required flag?

File: actions/jq.action.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    description: "JSON to query on"
+    required: true
+  query:
+    type: "string"
+    description: "Query to run on JSON"
+    required: true
+  unquoted:
+    type: "bool"
+    description: "If the final result is a string, it will be printed without quotes. [$JP_UNQUOTED]"
+    required: false
```
Comment (rimar-blink) [inline actions/jq.action.yaml:16]: it's not clear from the description what's the difference between jp and jq

File: actions/http_post.action.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Describes the action and it's parameters
+name: "http_post"
+description: "Executes post request on provided url"
+enabled: true
+parameters:
+  url:
+    type: "string"
+    description: "The url to send a post request"
+    required: true
+  contentType:
```
Comment (rimar-blink) [inline actions/http_post.action.yaml:10]: default value application/json?

File: actions/jq.action.yaml (hunk 4 above)
Comment (rimar-blink) [reply actions/jq.action.yaml:16]: unquoted field, what's the default?

File: implementation/http.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+	responseBytes, err := sendRequest(http.MethodPut, url, bodyAsString)
+	if err != nil {
+		return nil, err
+	}
+
+	return responseBytes, nil
+}
+
+func executeCoreHTTPDeleteAction(_ *plugin.ActionContext, request *plugin.ExecuteActionRequest) ([]byte, error) {
```
Comment (rimar-blink) [inline implementation/http.go:176]: all http methods should be handled by 1 or 2 functions, perhaps 1 for GET/HEAD and another one for all the rest that can have a body

File: implementation/http.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	body, err := readBody(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &httpResponse{
+		Status:           response.Status,
+		StatusCode:       response.StatusCode,
+		Proto:            response.Proto,
+		ProtoMajor:       response.ProtoMajor,
```
Comment (rimar-blink) [inline implementation/http.go:67]: ProtoMajor and ProtoMinor seem unneeded, please remove

File: implementation/http.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	respBytes, err := json.Marshal(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return respBytes, err
+}
+
+func sendRequest(method string, urlAsString string, data interface{}) ([]byte, error) {
+	postBody, err := json.Marshal(data)
```
Comment (rimar-blink) [inline implementation/http.go:86]: why json.Marshal?

File: implementation/implementation.go [hunk 9]
```diff
@@ -11,6 +11,11 @@ import (
 	"path"
 )
 
+type FinalOutput struct {
```
Comment (rimar-blink) [inline implementation/implementation.go:14]: What's so final about this output? :) Maybe call it CommandOutput?

File: implementation/ps.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	if !ok {
+		flags = ""
+	}
+	command := exec.Command("/bin/ps", flags)
+
+	outputBytes, execErr := command.CombinedOutput()
+	if execErr != nil {
+		log.Error("Detected failure, building result! Error: ", execErr)
+
+		failureResult := FinalOutput{Output: string(outputBytes), Error: execErr.Error()}
```
Comment (rimar-blink) [inline implementation/ps.go:21]: looks like the same code as in bash, jq, jp commands, can you refactor and extract a shared function?

**Review by rimar-blink — CHANGES_REQUESTED** [review]

File: actions/http_post.action.yaml (hunk 3 above)
Comment (itsofirblink) [reply actions/http_post.action.yaml:17]: Changed

File: actions/jq.action.yaml (hunk 4 above)
Comment (itsofirblink) [reply actions/jq.action.yaml:16]: Changed

File: actions/http_post.action.yaml (hunk 5 above)
Comment (itsofirblink) [reply actions/http_post.action.yaml:10]: Why?

File: implementation/http.go (hunk 7 above)
Comment (itsofirblink) [reply implementation/http.go:67]: Done

File: implementation/http.go (hunk 8 above)
Comment (itsofirblink) [reply implementation/http.go:86]: That's how you send data

File: implementation/implementation.go (hunk 9 above)
Comment (itsofirblink) [reply implementation/implementation.go:14]: Changed

File: implementation/http.go (hunk 6 above)
Comment (itsofirblink) [reply implementation/http.go:176]: why? it looks better like that

File: actions/http_delete.action.yaml (hunk 1 above)
Comment (itsofirblink) [reply actions/http_delete.action.yaml:1]: Why not?

File: actions/http_get.action.yaml (hunk 2 above)
Comment (itsofirblink) [reply actions/http_get.action.yaml:1]: Will add to post method

File: implementation/http.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	requestBody := bytes.NewBuffer(postBody)
+
+	// Create new http client with predefined options
+	client := &http.Client{
+		Timeout: time.Second * 60,
```
Comment (danielsblinkops) [inline implementation/http.go:91]: Looks like this should be configurable in tthe action

File: implementation/ps.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
+package implementation
+
+import (
+	"encoding/json"
+	"github.com/blinkops/blink-sdk/plugin"
+	log "github.com/sirupsen/logrus"
+	"os/exec"
+)
+
+func executeCorePsAction(_ *plugin.ActionContext, request *plugin.ExecuteActionRequest) ([]byte, error) {
```
Comment (danielsblinkops) [inline implementation/ps.go:10]: i really dont see a reason for this action to exit, ps inside a blink controlled docker container is not really interesting

File: implementation/implementation.go [hunk 13]
```diff
@@ -76,7 +81,15 @@ func NewCorePlugin(rootPluginDirectory string) (*CorePlugin, error) {
 	}
 
 	supportedActions := map[string]ActionHandler{
-		"python": executeCorePythonAction,
+		"python":      executeCorePythonAction,
```
Comment (danielsblinkops) [inline implementation/implementation.go:84]: lets change the python action to return only the output if there is no error

**Review by danielsblinkops — CHANGES_REQUESTED** [review]

File: implementation/json.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+	if !ok {
+		return nil, errors.New("no json provider for execution")
+	}
+
+	query, ok := request.Parameters[userProviderQueryKey]
+	if !ok {
+		return nil, errors.New("no query provider for execution")
+	}
+
+	command := exec.Command("/bin/echo", providedJson, "|", "/bin/jq", query)
```
[the remark this answers is not in the document]
Comment (itsofirblink) [reply implementation/json.go]: Done it in another way

File: actions/http_delete.action.yaml (hunk 1 above)
Comment (rimar-blink) [reply actions/http_delete.action.yaml:1]: it should be 1 http action, both from user perspective and from the code implementing it.

File: actions/http_get.action.yaml (hunk 2 above)
Comment (rimar-blink) [reply actions/http_get.action.yaml:1]: you should add headers to all http calls, one more reason to make it 1 action.

File: actions/http_post.action.yaml (hunk 5 above)
Comment (rimar-blink) [reply actions/http_post.action.yaml:10]: user convinience? we should ask liran about it.

File: implementation/http.go (hunk 8 above)
Comment (rimar-blink) [reply implementation/http.go:82]: what if it's not json?

File: implementation/http.go (hunk 6 above)
Comment (rimar-blink) [reply implementation/http.go:172]: not really

File: implementation/ps.go (hunk 12 above)
Comment (rimar-blink) [reply implementation/ps.go:10]: +1

**Review by rimar-blink — CHANGES_REQUESTED** [review]

**Review by danielsblinkops — APPROVED** [review]

**Pull request merged**